### DISCUSSION TITLE
Iterative scene parameter updates | creates | deletes

### DISCUSSION THREAD

**Discussion by georgy95** [mitsuba-renderer/mitsuba3#873] [Q&A] [answered]
I would like to update the scene parameters, without creating a new scene from scratch. 

It appears that once the scene is loaded, we can only `update` the parameters. A full set of good to have operations :
* Add `create` or `append` an object e.g. `Mesh` to the instantiated scene
* I would like to `delete` already instantiated scene objects (I though this is what `keep` was for)
* I would like to `overwrite` objects (logic below)

In the screenshot below, the `keep()` call does not propagate updates to the scene itself.
<img width="786" alt="screen" src="https://github.com/mitsuba-renderer/mitsuba3/assets/6961787/287853d9-3c85-4ada-b432-9b6276227939">

As for overwrites:
If my original scene `scene = mi.load_dict(mi.cornell_box())` had an root object with key `green`, I can overwrite it by supplying a new value e.g.

```
newgreen = {
'type': 'diffuse',
 'reflectance': {'type': 'rgb', 'value': [0.105421, 0.37798, 0.076425]}
}
params["green"] = mi.load_dict(newgreen) # overwrite or create
```

**Comment by njroussel** [comment, marked as answer]
Hi @georgy95 

Currently, the only modifications that can be applied to an instantiated scene are updates to certain plugin parameters, [as described in the tutorial](https://mitsuba.readthedocs.io/en/stable/src/rendering/editing_a_scene.html).

The `SceneParameters.keep()` method is practically syntactic sugar to programmatically identify certain parameters in your scene. It indeed, is not meant to delete objects.

The general recommendation is to instantiate objects (materials, integrators, geometry, ...) separately and only tie them together into a scene when necessary, if you want to avoid reloading a scene from scratch on every "major" update. By "major" update I mean:

* Adding/removing plugins
* Chaning the type of an existing plugin
* Modifying some plugin parameter that is not exposed through `mi.traverse()`

**Reply by georgy95** [reply]
It's possible to instantiate `integrators` and `materials` on their own and use them. For geometry-shapes, since they reference materials, can't be instantiated into a Mitsuba object on their own (due to material not found error).

Unless you meant to instantiate dictionary objects only. In which case, it would be almost equivalent to building the new scene from scratch.

**Reply by njroussel** [reply]
Indeed, you are correct. I hadn't really considered nested plugins, where the enclosing plugin would need to be re-created.

**Reply by georgy95** [reply]
Thanks! @njroussel

**Comment by gioarma** [comment]
I think that having the ability to add and delete objects would be nice. In my case, I use the Blender plugin to create a base `.xml` scene. Then, I would like to programmatically add other elements to the scene.
A workaround would be to convert the loaded xml file to a dictionary, modify the dictionary adding the objects, and then using `mi.load_dict`. However, looking at the documentation I see that there is no function to do that, I only see `mi.xml.dict_to_xml`, which does the opposite.

**Reply by njroussel** [reply]
A "better" workaround is to do what you suggest the other way around: keep the original scene description in XML, convert any additional elements you might want to add to your scene into XML (using `mi.xml.dict_to_xml` for example), and finally insert these new XML elements into the main scene XML which you finally load with `mi.load_file()` or `mi.load_string()`.

Of course, directly modifying a partially-loaded scene would still be better that re-loading new ones just to add or remove a single object. But that isn't available yet.

**Reply by gioarma** [reply]
Hi @njroussel, thanks for your answer. By "insert these new XML elements into the main scene" you mean programmatically opening the main XML and appending the new XML content, or is there another way of doing it that I'm not aware of?

**Reply by njroussel** [reply]
>you mean programmatically opening the main XML and appending the new XML content

Yes :smile: . Either just loading the file as a string and editing it or using whatever XML parsing/editor library you prefer.

**Reply by georgy95** [reply]
Maybe what I did will help you @gioarma 

Create a class to manage all of the scene assets and use the observer pattern to propagate the child component updates all the way up-to the parent. 

So when Texture changes, you update the Material, which then updates the Scene. 

Scene rebuilds took around 0.5s for me.
